llvm/lib/Target/AMDGPU/GCNHazardRecognizer.cpp

```diff
@@ -1040,6 +1096,37 @@ int GCNHazardRecognizer::checkInlineAsmHazards(MachineInstr *IA) {
     if (Op.isReg() && Op.isDef()) {
       WaitStatesNeeded =
           std::max(WaitStatesNeeded, checkVALUHazardsHelper(Op, MRI));
+
+      if (!TRI.isVectorRegister(MRI, Op.getReg()))
+        continue;
+
+      if (ST.hasDstSelForwardingHazard()) {
```

Comment:
Is this redundant with the early exit above? How does this interact with ST.has12DWordStoreHazard, which seems to be ignored

Reply:
In has12DWordStoreHazard case, WaitStatesNeeded = std::max(WaitStatesNeeded, checkVALUHazardsHelper(Op, MRI)); .  If hasDstSelForwardingHazard then we also account for possible waits needed for dest forwarding hazard.

I've touched up the code to make this more obvious and less reliant on early exit.